src/supervisor.tex

```diff
@@ -2261,3 +2281,108 @@ \chapter{``Svnapot'' Standard Extension for NAPOT Translation Contiguity, Versio
+be evicted from the cache by the latter.  If on the other hand the cacheable
+access appears after the non-cacheable access, then the former may remain
+cached as it normally would.
+
+Likewise, accesses performed under memory indicating the non-idempotent
+attribute must not be merged with idempotent accesses to the same region
+in flight at the same time, as the non-idempotency of the former must
+be respected.  This is not expected to be a common situation.
+
+Note that Svpbmt cannot be used to completely prevent speculative reads from
+being performed to a region of memory for which the PMAs indicate idempotence,
+as speculation can still be performed via M-mode or via Bare mappings, which do
+not use the PBMTs.
+\end{commentary}
+
+Memory accesses are ordered according to the effective memory type for the
+address in question.  A region of memory which is considered main memory by the
+PMAs but I/O by a PTE will obey channel 0 strong ordering, where for memory
+ordering purposes the ``I/O region'' is considered to be main memory.  In other
+words, each access to such a virtual address will appear in the global memory
+order after all prior operations from the same hart to main memory or to the
+same ``I/O region'', and it will appear earlier in the global memory order than
+all subsequent accesses from the same hart to main memory or to the same ``I/O
+region''.  Such operations are not guaranteed to remain ordered with respect to
+I/O operations to other I/O regions unless FENCEs are inserted by software.
+
+Likewise, a region of memory which is considered I/O by the PMAs but main
+memory by a PTE will obey RVWMO (or RVTSO if enabled) rather than I/O
+strong ordering rules, and accesses to such pages are considered main
+memory rather than I/O for the purposes of FENCE, {\em.aq}, and {\em.rl}.
+In such cases it is the responsibility of software to insert FENCEs
+or other ordering mechanisms if necessary.
+
+\begin{commentary}
+A device driver written to rely on I/O strong ordering rules will not
+operate correctly if the address range is mapped as main memory by the
+page-based memory types.  Operating systems and hypervisors must take
+care to apply such a combination only when strong ordering is not
+actually needed.
+
+In spite of the caveat above, it will often still be useful to map
+device memory regions in I/O as main memory so that write combining
+and speculative accesses can be performed, as such optimizations will
+likely improve performance when applied with adequate care.
+\end{commentary}
+
+The coherence PMA is not affected by Svpbmt.  Non-coherent I/O address ranges
+re-mapped by PBMTs into main memory must still rely on platform-specific
+mechanisms to enforce coherence with respect to external devices.  However,
+RVWMO and RVTSO still require eventual visibility of writes from one harts to
+other harts in the system.
+
+When two-level paging is enabled within the H extension, the page-based memory
```

Comment:
This text doesn't spell out how one combines the pbmt from VS-stage and G-stage PTEs when both are specifying overrides, i.e. which one wins out.

If one had names for the pbmt encodings, one could say something like:

_The combined PTE memory type is the more conservative of the VS-stage and G-stage memory types, i.e. IO wins out over NC which wins out over PMA.  The result is always one of IO, NC, or PMA._

Reply:
Conceptually I would view the PMA and G-stage PTE attributes being combined first.  This intermediate set of attributes then represents what is seen by a VM as PMA attributes - which the guest OS can then choose to condiitonally override.

This also provides an easy solution to specifying how VS-stage and G-stage pbmt's are combined.  One applies the earlier description - for how a PTE modifies PMA attributes - first with the G-stage pbmt, and then take that and apply the VS-stage pbmt.

Reply:
I don't understand...isn't the note about first/second exactly what the text below says?